### PR TITLE
Add WebGPU Emscripten platform/driver/samples.

### PR DESCRIPTION
**Pull request by ScottTodd**
Rollup of fixes (some simple, a few hacky as we figure out whether we want to continue supporting native WebGPU platforms), then the actual Emscripten platform code and samples.

The sample is forked from [`experimental/web/sample_dynamic/`](https://github.com/iree-org/iree/tree/main/experimental/web/sample_dynamic), but uses WebGPU instead of WebAssembly with Web Workers.

Both the new sample and the tests at [`experimental/web/testing/`](https://github.com/iree-org/iree/tree/main/experimental/web/testing) can successfully create WebGPU devices using Emscripten on my machine/browser (Chrome Canary on Windows with `#enable-unsafe-webgpu`). The sample and tests don't get much further than device creation yet (needs async invocations and readback).

The tests make use of [Asyncify](https://emscripten.org/docs/porting/asyncify.html) to implement synchronous versions of `wgpuInstanceRequestAdapter` and `wgpuAdapterRequestDevice`, while the sample preinitializes a `WGPUDevice` in JS then passes it in via Emscripten's `preinitializedWebGPUDevice`/`emscripten_webgpu_get_device()`. There are a few other ways we can set that up - we'll definitely want to disable Asyncify for shipping application code.

### CODE REVIEW

File: experimental/web/sample_webgpu/main.c [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+
+  if (!iree_status_is_ok(status)) {
+    iree_string_builder_deinitialize(&outputs_builder);
+    iree_status_fprint(stderr, status);
+    iree_status_free(status);
+    return "";
+  }
+
+  // Note: this leaks the buffer. It's up to the caller to free it after use.
+  return iree_string_builder_buffer(&outputs_builder);
```
Comment (benvanik) [inline experimental/web/sample_webgpu/main.c:439]: this is really shady - I know it's a sample but I think it's worth fixing this here before it proliferates/we forget :)

if you returned `strdup(iree_string_builder_buffer(...))` and then deinitialized the builder the caller could then call _free to free it when done

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+  const deviceWrapper = {queueId : WebGPU.mgrQueue.create(device["queue"])};
+  const deviceId = WebGPU.mgrDevice.create(device, deviceWrapper);
+  return deviceId;
+});
+
+#else
+
+WGPUAdapter wgpuInstanceRequestAdapterSync() {
+  fprintf(stderr, "wgpuInstanceRequestAdapterSync requires -sASYNCIFY\n");
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c:82]: is this the only place (so far) that needs asyncify? could have a global device that we query here (as mentioned in the comment above) but probably easiest to wait until we have everything working and then try to nuke asyncify altogether

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    return iree_make_status(
+        IREE_STATUS_UNAVAILABLE,
+        "WebGPU implementation not present or failed to load");
+  }
+
+  // Request an adapter from the implementation. We only get one of these and it
+  // may expose multiple devices so it's effectively what we consider a driver.
+  // HACKS: sync via Asyncify, assume success
+  WGPUAdapter adapter = wgpuInstanceRequestAdapterSync();
+  IREE_ASSERT_NE(adapter, NULL);
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c:154]: worth adding a check here instead - having webgpu doesn't meant you have an adapter, AFAIK (though maybe browsers hide webgpu entirely if no adapters would be returned - seems like that'd be expensive though)

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+static iree_status_t iree_hal_webgpu_emscripten_driver_create_device_by_id(
+    iree_hal_driver_t* base_driver, iree_hal_device_id_t device_id,
+    iree_host_size_t param_count, const iree_string_pair_t* params,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device) {
+  iree_hal_webgpu_emscripten_driver_t* driver =
+      iree_hal_webgpu_emscripten_driver_cast(base_driver);
+
+  // HACKS: sync via Asyncify, assume success
+  WGPUDevice device = wgpuAdapterRequestDeviceSync(driver->adapter);
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c:218]: also check here

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_util.c [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+#define WGPU_EMSCRIPTEN_INSTANCE ((WGPUInstance)((uintptr_t)0xABADD00Du))
+
+WGPUInstance wgpuCreateInstance(WGPUInstanceDescriptor const* descriptor) {
+  // Emscripten does not have instances (yet?)
+  // We use a sentinel value here so that we can do null checks in places for
+  // implementations that do use instances.
+  return WGPU_EMSCRIPTEN_INSTANCE;
+}
+
+void iree_wgpuBindGroupDrop(WGPUBindGroup bindGroup) {
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_util.c:22]: do the latest headers have the destroys for some of these? I think there's a buffer and device destroy now we can use (https://www.w3.org/TR/webgpu/#dom-gpudevice-destroy + https://www.w3.org/TR/webgpu/#dom-gpubuffer-destroy)

**Review by benvanik — CHANGES_REQUESTED** [review]
minor tweaks to handle errors but otherwise lgtm!

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c (hunk 2 above)
Comment (ScottTodd) [reply runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c:82]: Yep. Also worth calling out is "mapSync", which I couldn't find a browser-friendly solution for (didn't look that deeply though)

```c++
IREEWGPUBufferMapSyncStatus iree_wgpuBufferMapSync(WGPUDevice device,
                                                   WGPUBuffer buffer,
                                                   WGPUMapModeFlags mode,
                                                   size_t offset, size_t size) {
  IREEWGPUBufferMapSyncStatus status = IREEWGPUBufferMapSyncStatus_Unknown;
  wgpuBufferMapAsync(buffer, mode, offset, size,
                     iree_hal_webgpu_buffer_map_sync_callback, &status);
  // TODO(scotttodd): poll / wait somehow, or implement sync mapping differently
  // wgpuDevicePoll(device, /*force_wait=*/true);
  return status;
}
```

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_util.c (hunk 5 above)
Comment (ScottTodd) [reply runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_util.c:22]: For some (buffer, texture, queryset, device) yes.

* https://github.com/emscripten-core/emscripten/blob/704218f6a31dd5d2d584c5ec117b09ca424ffd0e/src/library_webgpu.js#L672-L693
* https://github.com/webgpu-native/webgpu-headers/blob/main/webgpu.h

Comment (benvanik) [reply runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_util.c:22]: Cool - we can clean those up later - not sure what we're supposed to do with other things though (still a massive oversight for native webgpu IMO, but I'm not sure how it doesn't leak with emscripten either!)

File: runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c (hunk 2 above)
Comment (benvanik) [reply runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c:82]: Yeah that'll be the big one - until we can remove that we'll probably be stuck with asyncify ;(

File: experimental/web/sample_webgpu/main.c (hunk 1 above)
Comment (ScottTodd) [reply experimental/web/sample_webgpu/main.c]: Thanks, I fixed that leak and added a link with some context.

File: runtime/src/iree/hal/buffer.c [hunk 6]
```diff
@@ -280,7 +280,10 @@ IREE_API_EXPORT iree_status_t iree_hal_buffer_validate_memory_type(
 IREE_API_EXPORT iree_status_t iree_hal_buffer_validate_access(
     iree_hal_memory_access_t allowed_memory_access,
     iree_hal_memory_access_t required_memory_access) {
-  if (iree_all_bits_set(allowed_memory_access, IREE_HAL_MEMORY_ACCESS_ANY)) {
+  // TODO(scotttodd): change to allowed_memory_access when possible
+  //   (at least the web samples aren't setting the required access bits)
+  // if (iree_all_bits_set(allowed_memory_access, IREE_HAL_MEMORY_ACCESS_ANY)) {
+  if (iree_all_bits_set(required_memory_access, IREE_HAL_MEMORY_ACCESS_ANY)) {
```
Comment (ScottTodd) [inline runtime/src/iree/hal/buffer.c:286]: I saw the `"memory access must specify one or more of _READ or _WRITE` error just below this running https://github.com/iree-org/iree/tree/main/experimental/web/sample_dynamic (CPU code), but I also see this error running with WebGPU:

`iree_api_webgpu.js:76 (C) iree/runtime/src/iree/hal/command_buffer_validation.c:80: PERMISSION_DENIED; requested buffer usage is not supported for the buffer on this queue; buffer allows TRANSFER|MAPPING, operation requires TRANSFER (allocator compatibility mismatch)`

Both will need to be fixed somehow?

File: runtime/src/iree/hal/drivers/webgpu/executable.c [hunk 7]
```diff
@@ -111,7 +111,7 @@ static iree_status_t iree_hal_webgpu_create_wgsl_shader_module(
               .next = NULL,
               .sType = WGPUSType_ShaderModuleWGSLDescriptor,
           },
-      .code = code,
+      .source = code,
```
Comment (ScottTodd) [inline runtime/src/iree/hal/drivers/webgpu/executable.c:114]: Note that this sort of change breaks the native build, until `IREE_DAWN_TAG`, `third_party/webgpu_headers/`, etc. are updated to a compatible commit. How much do we care to support native and browser builds?

**Pull request merged**